transport.go

```diff
@@ -32,27 +32,73 @@ var quicConfig = &quic.Config{
 }
 
 type connManager struct {
-	connIPv4Once sync.Once
-	connIPv4     net.PacketConn
+	mu sync.Mutex
+	// underhood PacketConn to its (re)use count
+	ipv4Conns map[net.PacketConn]int
+	ipv6Conns map[net.PacketConn]int
+}
 
-	connIPv6Once sync.Once
-	connIPv6     net.PacketConn
+func newConnManager() *connManager {
+	return &connManager{
+		ipv4Conns: make(map[net.PacketConn]int),
+		ipv6Conns: make(map[net.PacketConn]int),
+	}
 }
 
-func (c *connManager) GetConnForAddr(network string) (net.PacketConn, error) {
+func pickDialConn(conns map[net.PacketConn]int, remoteHostIP net.IP) net.PacketConn {
+	backup := make([]net.PacketConn, 0, len(conns))
+	for conn := range conns {
+		addr := conn.LocalAddr().(*net.UDPAddr)
+		if addr.IP.IsLoopback() && remoteHostIP.IsGlobalUnicast() {
+			continue
+		}
+		if addr.IP.IsLoopback() && remoteHostIP.IsLoopback() {
+			return conn
+		}
+		if (addr.IP.IsUnspecified() || addr.IP.IsGlobalUnicast()) && remoteHostIP.IsGlobalUnicast() {
+			return conn
+		}
+		if (addr.IP.IsUnspecified() || addr.IP.IsGlobalUnicast()) && remoteHostIP.IsLoopback() {
+			backup = append(backup, conn)
+		}
+	}
+	if len(backup) > 0 {
+		return backup[0]
+	}
+	return nil
+}
+
+func (c *connManager) GetConnForAddr(network, host string) (net.PacketConn, error) {
+	remoteAddr, err := net.ResolveUDPAddr(network, host)
+	if err != nil {
+		return nil, err
+	}
+
 	switch network {
 	case "udp4":
-		var err error
-		c.connIPv4Once.Do(func() {
-			c.connIPv4, err = c.createConn(network, "0.0.0.0:0")
-		})
-		return c.connIPv4, err
+		conn := pickDialConn(c.ipv4Conns, remoteAddr.IP)
+		if conn == nil {
+			conn, err = c.createConn(network, "0.0.0.0:0")
+			if err != nil {
+				return nil, err
+			}
+		}
+		c.mu.Lock()
+		c.ipv4Conns[conn] += 1
```

Comment:
What are we using these refcounts for?

Reply:
My initial thought was not actually close a socket when upper layer call `Close()`, except the refcount down to 0. But `go-reuseport-transport` doesn't do this, so I'd like to discuss if it's reasonable.